### PR TITLE
add an ellipsis if the cell content is too long

### PR DESCRIPTION
**Pull request by jochenberger**
I started to implement the ellipsis thing I asked for in #184. This is not complete yet, it does not work for the column that the tree is in and it fails for the last column in the `treegrid.html` example file.

### CODE REVIEW

**Comment by jochenberger** [comment]
See http://jsfiddle.net/tsvr92ge/2/

**Comment by deitch** [comment]
I love a PR that starts with "DO NOT MERGE"! :-)

I think that once we get it working on all columns except the tree, we should accept and merge. The tree itself is a different issue.

**Comment by deitch** [comment]
Either way, I will wait until you change the title to "MERGE!"

**Comment by jochenberger** [comment]
Yes, I need to figure out why it doesn't work in the example. It seems to work fine in the fiddle. Maybe I can find out something tomorrow.

**Comment by jochenberger** [comment]
The ellipsis should probably also added to the header cell.

**Comment by jochenberger** [comment]
It does actually does work with the last column in the example file, the content is just so short that the ellipsis itself would be too wide when the column is narrow enough that the content does not fit in. Quite hard to explain.
It should work for all columns except `treecol` now.
http://jsfiddle.net/tsvr92ge/6/

File: README.md [hunk 1]
```diff
@@ -81,6 +81,7 @@ We use the div to control the entire height and width, and the span to get acces
 * `columns`: an array of columns to create, on order. Each entry is an object with the following parameters:
 	* `tree`: boolean, whether the jstree should be placed in this column. Only the first `true` is accepted. If no column is set to `tree:true`, then the first column is used.
 	* `width`: width of the column in pixels. If no width is given, the default is `auto` **except for the last column**. In the last column, if no width is given, it is treated as 'auto' and fills the entire rest of the grid to the right.
+	* `ellipsis`: true/false. If true (the default), then, if the column content overflows the column width, an ellipsis (…) is shown at the end.
```
Comment (deitch) [inline README.md]: Based on the commit that vakata did, I was thinking:

* `true`: always show an ellipsis if too long (i.e. apply the class)
* `false`: never show an ellipsis if too long (current default)
* `undefined` (in other words, user does not pick anything): follow whatever the rule is for the tree itself.

Thoughts?

File: jstreegrid.js [hunk 2]
```diff
@@ -534,6 +534,9 @@
 				col = this.midWrapper.children("div.jstree-grid-column-"+i);
 				last = $("<div></div>").css(conf).addClass("jstree-grid-div-"+this.uniq+"-"+i+" "+(tr?"ui-widget-header ":"")+" jstree-grid-header jstree-grid-header-cell jstree-grid-header-"+classAdd+" "+cl+" "+ccl).html(val);
 				last.addClass((tr?"ui-widget-header ":"")+"jstree-grid-header jstree-grid-header-"+classAdd);
+				if (col.ellipsis !== false){
```
Comment (deitch) [inline jstreegrid.js]: Following the comments on the issue:

````javascript
if (col.ellipsis === true || (col.ellipsis === undefined && this.settings.themes.ellipsis === true)) {
````

(I might have the above wrong in terms of retrieving the setting, but the idea)

File: jstreegrid.js [hunk 3]
```diff
@@ -966,6 +969,10 @@
 						} else {
 							last.removeClass(SEARCHCLASS);
 						}
+						// TODO this does not work for the tree column yet
+						if (i !== treecol && col.ellipsis !== false) {
```
Comment (deitch) [inline jstreegrid.js]:
````javascript
if (I !== treecol && (col.ellipsis === true || (col.ellipsis === undefined && this.settings.themes.ellipsis === true))) {
````

(I might have the above wrong in terms of retrieving the setting, but the idea)

**Comment by deitch** [comment]
Added some comments in the code of this commit.

Also, two more things:

1. Don't we want to support grid-wide, not just per-column?
2. We would need to capture the events.

Events (around L450) (again, pseudo-code):

````javascript
.on('show_ellipsis.jstree', $.proxy(function(e,data){
    if (this.settings.ellipsis === undefined) {
      this.gridWrapper.find(".jstree-grid-cell").addClass("jstree-grid-ellipsis");
    }
},this))
.on('hide_ellipsis.jstree', $.proxy(function(e,data){
    if (this.settings.ellipsis === undefined) {
      this.gridWrapper.find(".jstree-grid-cell").removeClass("jstree-grid-ellipsis");
    }
},this))
````

File: README.md (hunk 1 above)
Comment (jochenberger) [inline README.md]: Maybe we should leave this out and only render the ellipsis based on the jsTree setting for now. We can always add the setting for the plugin later.

**Comment by jochenberger** [comment]
See http://jsfiddle.net/tsvr92ge/7/

**Comment by deitch** [comment]
@jochenberger I don't understand. The change (and implementation in jsfiddle) seems to:

1. Remove treegrid-specific setting in favour of inheriting from core tree setting
2. Eliminate ability to change along with core tree

I thought you wanted to have the ability to do any of:

* grid always on, independent of tree
* grid always off, independent of tree
* grid always in sync with tree

**Comment by jochenberger** [comment]
Well, all *I* want is to show the ellipsis if the content is too long, for the tree as well as for the grid. :-)
I *may* want to be able to control the behavior on a per-column basis, but I'm not sure about that and I can always create another PR to address this later.

**Comment by deitch** [comment]
>  I can always create another PR to address this later.

Now I *know* that @jochenberger is my kind of people. Fix one thing, then the next, then the next. Don't try to "boil the ocean". Love it.

(and by the way, I don't really like per-column ellipsis settings personally myself, but I am just one person)

So, how about we take what there is now and add the dynamic watching of events. We merge that in, then add another PR to allow separate settings grid vs tree? To get there, all we would need for this is the event watchers:

````js
.on('show_ellipsis.jstree', $.proxy(function(e,data){
      this.gridWrapper.find(".jstree-grid-cell").addClass("jstree-grid-ellipsis");
},this))
.on('hide_ellipsis.jstree', $.proxy(function(e,data){
      this.gridWrapper.find(".jstree-grid-cell").removeClass("jstree-grid-ellipsis");
},this))
````

Then we can merge it right on in.

**Comment by jochenberger** [comment]
It already does the event thing. ;-)

**Comment by deitch** [comment]
I'd like to say, "I have no idea how I missed that," but "engineer's senility" is the most likely answer!

Then I think this LGTM. Any issues with merging?

**Comment by jochenberger** [comment]
> Any issues with merging?

None that I know of. @QueroBartK?

**Comment by QueroBartK** [comment]
This looks good to go if you ask me. :-)

**Comment by deitch** [comment]
Bumped version number to 3.5.29 (this should be a minor version number bump, but will wait for the other 2 PRs first) and `npm publish` done.

**Pull request merged**